x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/ProgressListenableActionFuture.java

```diff
@@ -0,0 +1,216 @@
+        ensureNotCompleted();
+
+        if (value < start) {
+            assert false : value + " < " + start;
+            throw new IllegalArgumentException("Cannot update progress with a value less than [start=" + start + ']');
+        }
+        if (end < value) {
+            assert false : end + " < " + value;
+            throw new IllegalArgumentException("Cannot update progress with a value greater than [end=" + end + ']');
+        }
+
+        List<ActionListener<Long>> listenersToExecute = null;
+        synchronized (this) {
+            assert progress == null || progress < value : progress + " < " + value;
+            this.progress = value;
+
+            final List<Tuple<Long, ActionListener<Long>>> listeners = this.listeners;
+            if (listeners != null) {
+                List<Tuple<Long, ActionListener<Long>>> listenersToKeep = null;
+                for (Tuple<Long, ActionListener<Long>> listener : listeners) {
+                    if (value < listener.v1()) {
+                        if (listenersToKeep == null) {
+                            listenersToKeep = new ArrayList<>();
+                        }
+                        listenersToKeep.add(listener);
+                    } else {
+                        if (listenersToExecute == null) {
+                            listenersToExecute = new ArrayList<>();
+                        }
+                        listenersToExecute.add(listener.v2());
+                    }
+                }
+                this.listeners = listenersToKeep;
+            }
+        }
+        if (listenersToExecute != null) {
+            listenersToExecute.forEach(listener -> executeListener(listener, () -> value));
+        }
+        assert invariant();
+    }
+
+    @Override
+    public void onResponse(Long result) {
+        ensureNotCompleted();
+        super.onResponse(result);
+    }
+
+    @Override
+    public void onFailure(Exception e) {
+        ensureNotCompleted();
+        super.onFailure(e);
+    }
+
+    private void ensureNotCompleted() {
+        if (completed) {
+            throw new IllegalStateException("Future is already completed");
+        }
+    }
+
+    @Override
+    protected void done() {
+        super.done();
+        final List<Tuple<Long, ActionListener<Long>>> listenersToExecute;
+        synchronized (this) {
+            completed = true;
+            listenersToExecute = this.listeners;
+            listeners = null;
+        }
+        if (listenersToExecute != null) {
+            listenersToExecute.stream().map(Tuple::v2).forEach(listener -> executeListener(listener, () -> actionGet(0L)));
+        }
+        assert invariant();
+    }
+
+    /**
+     * Attach a {@link ActionListener} to the current future. The listener will be executed once the future is completed.
+     *
+     * @param listener the {@link ActionListener} to add
+     */
+    @Override
+    public void addListener(final ActionListener<Long> listener) {
```

Comment:
Can we drop this (and the `implements ListenableActionFuture<Long>` that requires it)? I think callers should always specify their target endpoint.